### PR TITLE
Add command to list Java projects

### PR DESCRIPTION
**Pull request by jdneo**

### CODE REVIEW

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestItemUtils.java [hunk 1]
```diff
@@ -22,47 +22,53 @@
...
-
-    public static TestItem constructTestItem(IJavaElement element, TestLevel level, TestKind kind)
+    public static JavaTestItem constructJavaTestItem(IJavaElement element, TestLevel level, TestKind kind)
             throws JavaModelException {
         final String displayName;
         final String fullName;
         if (element instanceof IPackageFragment && ((IPackageFragment) element).isDefaultPackage()) {
             displayName = DEFAULT_PACKAGE_NAME;
-            fullName = DEFAULT_PACKAGE_NAME;
+            fullName = "";
```
Comment (Eskibear) [inline java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestItemUtils.java:41]: is empty string working?

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestItemUtils.java [hunk 2]
```diff
@@ -22,47 +22,53 @@
...
         if (element instanceof ISourceReference) {
-            // getSourceRange() is not used here because we want the Code Lens appear above the
-            // method name, not above the annotation.
-            final ISourceRange range = ((ISourceReference) element).getNameRange();
-            return JDTUtils.toRange(element.getOpenable(), range.getOffset(), range.getLength());
+            final ISourceRange sourceRange = ((ISourceReference) element).getSourceRange();
+            final ISourceRange nameRange = ((ISourceReference) element).getNameRange();
+            if (SourceRange.isAvailable(sourceRange) && SourceRange.isAvailable(nameRange)) {
+                return JDTUtils.toRange(element.getOpenable(), nameRange.getOffset(),
+                        sourceRange.getLength() - nameRange.getOffset() + sourceRange.getOffset());
```
Comment (Eskibear) [inline java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestItemUtils.java:65]: `sourceRange.getLength() - nameRange.getOffset() + sourceRange.getOffset()` equals to `nameRange.getLength()`?

can you briefly explain?

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestItemUtils.java (hunk 1 above)
Comment (jdneo) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestItemUtils.java:41]: Yes default package's element name is empty.

While here I should get the fullName in just one call, no need to handle separately, will update it.

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java [hunk 3]
```diff
@@ -11,17 +11,75 @@
...
+            }
+            if (project.getProject().equals(JavaLanguageServerPlugin.getProjectsManager().getDefaultProject())) {
+                continue;
+            }
+            if (workspaceFolderPath.isPrefixOf(project.getProject().getLocation())) {
+                javaProjects.add(project);
+                continue;
+            }
+
+            final IPath linkedFolderLocation = project.getProject()
```
Comment (Eskibear) [inline java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:63]: what's this part for? symlink?

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java [hunk 4]
```diff
@@ -11,17 +11,75 @@
...
+        }
+        final String workspaceFolderUri = (String) arguments.get(0);
+        final IPath workspaceFolderPath = ResourceUtils.filePathFromURI(workspaceFolderUri);
+        if (workspaceFolderPath == null) {
+            JUnitPlugin.logError("Failed to parse workspace folder path from uri: " + workspaceFolderUri);
+            // todo: handle non-file scheme
+            return Collections.emptyList();
+        }
+        final List<IJavaProject> javaProjects = new LinkedList<>();
+        for (final IJavaProject project : ProjectUtils.getJavaProjects()) {
```
Comment (Eskibear) [inline java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:51]: Among `ProjectUtils.getJavaProjects()`, what else are you going to filter together with default project?

**Review by Eskibear — COMMENTED** [review]
Overall the list projects part LGTM. But I remember JDTLS also has similar utility, is it exposed?

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestItemUtils.java (hunk 2 above)
Comment (jdneo) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestItemUtils.java:65]: sourceRange contains JavaDocs, so `(nameRange.getOffset() - sourceRange.getOffset())` equals to the length of JavaDoc, I'll add a comment for it.

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java (hunk 3 above)
Comment (jdneo) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:63]: invisible project

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java (hunk 4 above)
Comment (jdneo) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:51]: No other projects so far. All the projects that can be seen in user's workspaces should be listed

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java (hunk 3 above)
Comment (Eskibear) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:63]: got it.

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java (hunk 4 above)
Comment (Eskibear) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:51]: Just curious about if we can simply remove default project, and skip the "prefix checking"?

Comment (Eskibear) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:51]: prefix checking it totally fine with me, just curious about it.

**Comment by jdneo** [comment]
JDTLS has a utility to get all Java projects (ProjectUtils.getJavaProjects()), which I've used here

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java (hunk 4 above)
Comment (jdneo) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:51]: Oh, that's a good point. As far as I know, currently we could have default project, visible project and invisible project. It seems that we can just remove the default project and put other Java projects into the list.

**Review by Eskibear — APPROVED** [review]

**Pull request merged**